kubetest2-gke/deployer/build/gke_make.go

```
@@ -0,0 +1,108 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package build
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"k8s.io/klog"
+
+	"sigs.k8s.io/kubetest2/pkg/build"
+	"sigs.k8s.io/kubetest2/pkg/exec"
+)
+
+type gkeBuildAction string
+
+const (
+	compile      gkeBuildAction = "compile"
+	pack         gkeBuildAction = "package"
+	stage        gkeBuildAction = "push-gcs"
+	printVersion gkeBuildAction = "print-version"
+)
+
+const (
+	// GKEMakeStrategy builds and stages using the gke_make build
+	GKEMakeStrategy build.BuildAndStageStrategy = "gke_make"
+)
+
+type GKEMake struct {
+	RepoRoot      string
+	BuildScript   string
+	VersionSuffix string
+	StageLocation string
+}
+
+func gkeBuildActions(actions []gkeBuildAction) string {
+	stringActions := make([]string, len(actions))
+	for i, action := range actions {
+		stringActions[i] = string(action)
+	}
+	return strings.Join(stringActions, ",")
+}
+
+func arg(key, value string) string {
+	return fmt.Sprintf("%s=%s", key, value)
+}
+
+func (gmb *GKEMake) runWithActions(stdout, stderr io.Writer, actions []gkeBuildAction, extraArgs ...string) error {
+
+	args := []string{arg("GKE_BUILD_ACTIONS", gkeBuildActions(actions))}
+	args = append(args, extraArgs...)
+	cmd := exec.Command(gmb.BuildScript, args...)
+	cmd.SetDir(gmb.RepoRoot)
+	cmd.SetStdout(stdout)
+	cmd.SetStderr(stderr)
+	return cmd.Run()
+}
+
+func (gmb *GKEMake) Build() (string, error) {
+	klog.V(2).Infof("starting gke build ...")
+	// Skip validation for faster builds
+	// TODO: add support for a separate validate mode
+	if err := gmb.runWithActions(os.Stdout, os.Stderr, []gkeBuildAction{compile, pack}, arg("VERSION_SUFFIX", gmb.VersionSuffix)); err != nil {
+		return "", err
+	}
+	version := &bytes.Buffer{}
+	if err := gmb.runWithActions(version, ioutil.Discard, []gkeBuildAction{printVersion}, arg("VERSION_SUFFIX", gmb.VersionSuffix)); err != nil {
+		return "", err
+	}
```

suggest running this before compile/package so if it fails we fail fast

---

done